ding/reward_model/network.py

```diff
@@ -0,0 +1,428 @@
+            - next_state (:obj:`torch.Tensor`):
+                The next state batch
+            - action_long (:obj:`torch.Tensor`):
+                The action batch
+        Returns:
+            - real_next_state_feature (:obj:`torch.Tensor`):
+                Run with the encoder. Return the real next_state's embedded feature.
+            - pred_next_state_feature (:obj:`torch.Tensor`):
+                Run with the encoder and residual network. Return the predicted next_state's embedded feature.
+            - pred_action_logit (:obj:`torch.Tensor`):
+                Run with the encoder. Return the predicted action logit.
+        Shapes:
+            - state (:obj:`torch.Tensor`): :math:`(B, N)`, where B is the batch size and N is ''obs_shape''
+            - next_state (:obj:`torch.Tensor`): :math:`(B, N)`, where B is the batch size and N is ''obs_shape''
+            - action_long (:obj:`torch.Tensor`): :math:`(B)`, where B is the batch size''
+            - real_next_state_feature (:obj:`torch.Tensor`): :math:`(B, M)`, where B is the batch size
+              and M is embedded feature size
+            - pred_next_state_feature (:obj:`torch.Tensor`): :math:`(B, M)`, where B is the batch size
+              and M is embedded feature size
+            - pred_action_logit (:obj:`torch.Tensor`): :math:`(B, A)`, where B is the batch size
+              and A is the ''action_shape''
+        """
+        action = one_hot(action_long, num=self.action_shape)
+        encode_state = self.feature(state)
+        encode_next_state = self.feature(next_state)
+        # get pred action logit
+        concat_state = torch.cat((encode_state, encode_next_state), 1)
+        pred_action_logit = self.inverse_net(concat_state)
+        # ---------------------
+
+        # get pred next state
+        pred_next_state_feature_orig = torch.cat((encode_state, action), 1)
+        pred_next_state_feature_orig = self.forward_net_1(pred_next_state_feature_orig)
+
+        # residual
+        for i in range(4):
+            pred_next_state_feature = self.residual[i * 2](torch.cat((pred_next_state_feature_orig, action), 1))
+            pred_next_state_feature_orig = self.residual[i * 2 + 1](
+                torch.cat((pred_next_state_feature, action), 1)
+            ) + pred_next_state_feature_orig
+        pred_next_state_feature = self.forward_net_2(torch.cat((pred_next_state_feature_orig, action), 1))
+        real_next_state_feature = encode_next_state
+        return real_next_state_feature, pred_next_state_feature, pred_action_logit
+
+    def learn(self, state: torch.Tensor, next_state: torch.Tensor,
+              action_long: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, float]:
+        real_next_state_feature, pred_next_state_feature, pred_action_logit = self._forward(
+            state, next_state, action_long
+        )
+
+        inverse_loss = F.cross_entropy(pred_action_logit, action_long.long())
```

Comment:
action_long.long() -> action.long()

Reply:
it may be not a good idea, because action will used in other place